noether_tpp/src/tool_path_modifiers/organization_modifiers.cpp

```diff
@@ -57,4 +59,175 @@ ToolPaths SnakeOrganizationModifier::modify(ToolPaths tool_paths) const
+ * is as close as possible to the end of the previous tool path (or a reference position, in the case of the first tool
+ * path)
+ * @details The primary intent of this function is to organize tool path segments for edge tool paths; these segments
+ * represent segments of an edge and should be arranged in a sequential and connected order.
+ */
+static std::vector<std::size_t> sortToolPathBySegmentConnection(const ToolPath& path,
+                                                                const Eigen::Vector3d& start_reference)
+{
+  std::vector<std::size_t> segment_order;
+  segment_order.reserve(path.size());
+
+  // Comparator function for finding the segment whose first waypoint is closest to the reference position
+  auto compare_distances = [&path](const std::size_t& a, const std::size_t& b, const Eigen::Vector3d& ref) {
+    double d_a = (path.at(a).at(0).translation() - ref).norm();
+    double d_b = (path.at(b).at(0).translation() - ref).norm();
+    return d_a < d_b;
+  };
+
+  while (segment_order.size() < path.size())
+  {
+    // Get the indices of the yet unordered segments
+    std::vector<std::size_t> remaining_segment_indices;
+    {
+      std::vector<std::size_t> order(path.size());
+      std::iota(order.begin(), order.end(), 0);
+      std::set_difference(order.begin(),
+                          order.end(),
+                          segment_order.begin(),
+                          segment_order.end(),
+                          std::back_inserter(remaining_segment_indices));
+    }
+
+    if (remaining_segment_indices.size() == 1)
+    {
+      segment_order.push_back(remaining_segment_indices.front());
+    }
+    else
+    {
+      // Get the reference point with which to compare the segment start positions
+      Eigen::Vector3d ref;
+      if (segment_order.empty())
+      {
+        // Use the reference start position if no segments have been ordered yet
+        ref = start_reference;
+      }
+      else
+      {
+        // Otherwise use the position of the last waypoint in the last ordered segment
+        ref = path.at(segment_order.back()).back().translation();
+      }
+
+      // Find the segment whose first waypoint is closest to the reference position
+      auto comp = std::bind(compare_distances, std::placeholders::_1, std::placeholders::_2, ref);
+      std::sort(remaining_segment_indices.begin(), remaining_segment_indices.end(), comp);
+
+      // Add the index of that segment to the output variable
+      segment_order.push_back(remaining_segment_indices.front());
+    }
+  }
+
+  return segment_order;
+}
+
+/**
+ * @brief Helper function for creating a re-ordered copy of a vector
+ */
+template <typename T>
+static std::vector<T> reorder(const std::vector<T>& input, const std::vector<std::size_t>& order)
+{
+  std::vector<T> output;
+  output.reserve(input.size());
+
+  for (std::size_t idx : order)
+  {
+    output.push_back(input.at(idx));
```

Comment:
Any particular reason for using `.at(idx)` instead of `[idx]`?

Reply:
I prefer for this to throw an exception rather than produce a seg-fault